### ISSUE TITLE
OS X support

### ISSUE THREAD

**Issue by bunster** [howl-editor/howl#68]
I'm running on an OSX and during the make process, this error pops up. I've installed glib via Homebrew and still it calls that error. Anyone know what I can do to fix this?

**Comment by kirbyfan64** [comment]
What's the full error message? I'm pretty sure there has to be *something* from `pkg-config`...

**Comment by bunster** [comment]
[Here](http://pastebin.com/raw.php?i=mFZ9p7HS) is the paste of it from start-to-error.

**Comment by bunster** [comment]
I notice it said I didn't have gtk+-3.0, currently I have gtk+2.24.28. That may be part of the problem.

**Comment by kirbyfan64** [comment]
You need to install Gtk+ 3:

```
$ brew install gtk+3
```

You might also want to install `md5sum` to get rid of the checksum warnings:

```
brew install md5sha1sum
```

I read that OSX's `md5` command has an option `-r` that causes it to work the same way as `md5sum`. Should the Makefile use that on OSX?

**Comment by bunster** [comment]
Aye, I am just installing gtk+3 now :) And I installed md5sha1sum too. I'll try again once this is done and update the posts here!

**Comment by kirbyfan64** [comment]
I put together the following patch:

```diff
diff --git a/src/process_helpers.c b/src/process_helpers.c
index d71066e..efc33b3 100644
--- a/src/process_helpers.c
+++ b/src/process_helpers.c
@@ -25,7 +25,10 @@ int sig_USR2 = SIGUSR2;
 int sig_PIPE = SIGPIPE;
 int sig_ALRM = SIGALRM;
 int sig_TERM = SIGTERM;
+#ifdef __linux
 int sig_STKFLT = SIGSTKFLT;
+int sig_PWR = SIGPWR;
+#endif
 int sig_CHLD = SIGCHLD;
 int sig_CONT = SIGCONT;
 int sig_STOP = SIGSTOP;
@@ -38,6 +41,10 @@ int sig_XFSZ = SIGXFSZ;
 int sig_VTALRM = SIGVTALRM;
 int sig_PROF = SIGPROF;
 int sig_WINCH = SIGWINCH;
+#ifdef _DARWIN_C_SOURCE
+int sig_IOT = SIGIOT;
+int sig_EMT = SIGEMT;
+#else
 int sig_POLL = SIGPOLL;
-int sig_PWR = SIGPWR;
+#endif
 int sig_SYS = SIGSYS;
```

Save it into a file named `darwin_fix.patch` and run `git apply darwin_fix.patch`. If this works, I'll open a PR.

**Comment by kirbyfan64** [comment]
Related to https://github.com/luaposix/luaposix/issues/24.

**Comment by bunster** [comment]
Thanks! Should I run that patch in the src folder?

**Comment by kirbyfan64** [comment]
@bunster It should work in any folder in the `howl` directory. So, yes, you can run it in the `src` folder.

**Comment by bunster** [comment]
Looks like two of the errors went away, but there's still one left.
```
[23:31][boonie@bunhill][src (master *% u=)]$ make install
cc -c process_helpers.c -Wall -O2 -g -I/Users/boonie/Projects/howl/src/deps/LuaJIT-2.0.4/src -Ideps/scintilla342-howl-1/include -DGTK=1 -D_REENTRANT -I/usr/local/Cellar/gtk+3/3.16.6/include/gtk-3.0 -I/usr/local/Cellar/glib/2.44.1/include/gio-unix-2.0/ -I/usr/local/Cellar/cairo/1.14.2_1/include/cairo -I/usr/local/Cellar/libepoxy/1.3.1/include -I/usr/local/Cellar/pango/1.36.8_2/include/pango-1.0 -I/usr/local/Cellar/atk/2.16.0/include/atk-1.0 -I/usr/local/Cellar/cairo/1.14.2_1/include/cairo -I/usr/local/Cellar/pixman/0.32.6/include/pixman-1 -I/usr/local/Cellar/fontconfig/2.11.1/include -I/usr/local/Cellar/freetype/2.5.3_1/include/freetype2 -I/usr/local/Cellar/freetype/2.6_1/include/freetype2 -I/usr/local/Cellar/libpng/1.6.18/include/libpng16 -I/usr/local/Cellar/gdk-pixbuf/2.30.8/include/gdk-pixbuf-2.0 -I/usr/local/Cellar/libpng/1.6.18/include/libpng16 -I/usr/local/Cellar/glib/2.44.1/include/glib-2.0 -I/usr/local/Cellar/glib/2.44.1/lib/glib-2.0/include -I/usr/local/opt/gettext/include  -DHOWL_PREFIX=/usr/local
process_helpers.c:48:16: error: use of undeclared identifier 'SIGPOLL'
int sig_POLL = SIGPOLL;
               ^
1 error generated.
make: *** [process_helpers.o] Error 1
```

**Comment by kirbyfan64** [comment]
@bunster Try changing `_DARWIN_C_SOURCE` to `__APPLE__`.

**Comment by bunster** [comment]
Looks like that worked! The error went away, but another one has popped up at the very end of the make process:
```
==== Successfully built LuaJIT 2.0.4 ====
cc -o howl main.o process_helpers.o deps/lpeg-0.10.2/lpeg.o deps/LuaJIT-2.0.4/src/libluajit.a deps/scintilla342-howl-1/bin/scintilla.a -lm -ldl -L/usr/local/Cellar/gtk+3/3.16.6/lib -L/usr/local/Cellar/pango/1.36.8_2/lib -L/usr/local/Cellar/atk/2.16.0/lib -L/usr/local/Cellar/cairo/1.14.2_1/lib -L/usr/local/Cellar/gdk-pixbuf/2.30.8/lib -L/usr/local/Cellar/glib/2.44.1/lib -L/usr/local/opt/gettext/lib -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgmodule-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl  -lstdc++ -Wl,-E
ld: unknown option: -E
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [howl] Error 1
[01:36][boonie@bunhill][src (master *% u=)]$
```

**Comment by kirbyfan64** [comment]
Try editing `src/Makefile` and changing:

```make
LD_FLAGS = -Wl,-E
```

to:

```make
LD_FLAGS = -Wl,-export_dynamic
```

**Comment by nilnor** [comment]
I believe you might be the first one to attempt an OS X build that I know of, so I'm afraid there are some hurdles to overcome (I don't have a mac myself and so haven't made the attempt). You seem to be very close now though (thanks @kirbyfan64!), so I'm very interested to hear how it turns out.  It would be great to be able to add build instructions for OS X, along with the necessary changes @kirbyfan64 has dug up.

A final note: In addition to the `-export_dynamic` flag suggested, some other linker flags need to be added in order for LuaJIT to work correctly on 64 bit OS X. So please try the following for `LD_FLAGS`:

```
LD_FLAGS = -Wl,-export_dynamic -pagezero_size 10000 -image_base 100000000
```

**Comment by bunster** [comment]
I took a nap, back now though. Let's do this! :dancer:

**Comment by kirbyfan64** [comment]
Ah, my bad. Go back to `process_helpers.c`. Change this:

```c
#ifdef __linux
int sig_STKFLT = SIGSTKFLT;
int sig_PWR = SIGPWR;
#endif
```

to

```c
#ifndef __linux
#define SIGSTKFLT 0
#define SIGPWR 0
#endif
int sig_STKFLT = SIGSTKFLT;
int sig_PWR = SIGPWR;
```

and this:

```c
#ifdef __APPLE__
int sig_IOT = SIGIOT;
int sig_EMT = SIGEMT;
#else
int sig_POLL = SIGPOLL;
#endif
```

to:

```c
#ifdef __APPLE__
#define SIGPOLL 0
#else
#define SIGIOT 0
#define SIGEMT 0
#endif
int sig_IOT = SIGIOT;
int sig_EMT = SIGEMT;
int sig_POLL = SIGPOLL;
```

Now go to `lib/howl/io/process.moon` and change this line:

```moonscript
  'PWR', 'SYS'
```

to:

```moonscript
  'PWR', 'SYS', 'IOT', 'EMT'
```

and change this line:

```moonscript
  signals[s] = tonumber C["sig_#{s}"]
```

to:

```moonscript
  sig = tonumber C["sig_#{s}"]
  signals[s] = sig if sig != 0
```

**MAKE SURE THE INDENTATION FOR THE EDITS TO `lib/howl/process.moon` REMAINS THE SAME!!** The leading spaces are *very* important!

**Comment by nilnor** [comment]
Note that the change to `process.moon` can also be skipped to simplify, at least for now, as the new signals aren't actually used.

**Comment by kirbyfan64** [comment]
@nilnor The second change still needs to be done to avoid the `dlopen` error, though.

**Comment by bunster** [comment]
Do you mean the darwin_patch instead of process_helpers.c? And I notice your changes don't have the beginning + sign, does that mean I take those out and just have the #'s?

**Comment by nilnor** [comment]
@kirbyfan64 if the signals aren't listed in `process.moon` there will be no attempt to resolve them (referring to the `IOT`, `EMT` signals here).

**Comment by kirbyfan64** [comment]
@bunster I meant `process_helpers.c`. The full path is `src/process_helpers.c`.

**Comment by kirbyfan64** [comment]
@nilnor But OSX still doesn't have `SIGPOLL` and the Linux-only signals. So the change is still necessary.

**Comment by bunster** [comment]
My `process_helpers.c` file looks like this at the moment:
```
/* Copyright 2012-2015 The Howl Developers */
/* License: MIT (see LICENSE.md at the top-level directory of the distribution) */

#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>

int process_exited_normally(int status) { return WIFEXITED(status);  }
int process_exit_status(int status) { return WEXITSTATUS(status);  }
int process_was_signalled(int status) { return WIFSIGNALED(status);  }
int process_get_term_sig(int status) { return WTERMSIG(status);  }

int sig_HUP = SIGHUP;
int sig_INT = SIGINT;
int sig_QUIT = SIGQUIT;
int sig_ILL = SIGILL;
int sig_TRAP = SIGTRAP;
int sig_ABRT = SIGABRT;
int sig_BUS = SIGBUS;
int sig_FPE = SIGFPE;
int sig_KILL = SIGKILL;
int sig_USR1 = SIGUSR1;
int sig_SEGV = SIGSEGV;
int sig_USR2 = SIGUSR2;
int sig_PIPE = SIGPIPE;
int sig_ALRM = SIGALRM;
int sig_TERM = SIGTERM;
int sig_STKFLT = SIGSTKFLT;
int sig_CHLD = SIGCHLD;
int sig_CONT = SIGCONT;
int sig_STOP = SIGSTOP;
int sig_TSTP = SIGTSTP;
int sig_TTIN = SIGTTIN;
int sig_TTOU = SIGTTOU;
int sig_URG = SIGURG;
int sig_XCPU = SIGXCPU;
int sig_XFSZ = SIGXFSZ;
int sig_VTALRM = SIGVTALRM;
int sig_PROF = SIGPROF;
int sig_WINCH = SIGWINCH;
int sig_POLL = SIGPOLL;
int sig_PWR = SIGPWR;
int sig_SYS = SIGSYS;
```
There are no ifdefs. The ifdefs you're talking about are in the patch you gave me.

**Comment by nilnor** [comment]
@kirbyfan64 yes the change to `process_helpers.c` is necessary, but that should take care of the missing signals not being available for later resolving. Hence the `process.moon` shouldn't be strictly needed to proceed right now. 

@bunster Did you you a clean check-out? That's not now the `process_helpers.c` file would like after the previous patch/changes from @kirbyfan64.

**Comment by kirbyfan64** [comment]
@bunster That's weird... :/

The just go to the top of `process_helpers.c` and add this under `#include <signal.h>`:

```c
#ifndef __linux
#define SIGSTKFLT 0
#define SIGPWR 0
#endif

#ifdef __APPLE__
#define SIGPOLL 0
#endif
```

And still do the second change to `process.moon` I mentioned.

**Comment by bunster** [comment]
I made those changes but I'm running into the same problem when I try to start howl :(
```
./howl
/Users/boonie/Projects/howl/lib/howl/init.lua:64: /Users/boonie/Projects/howl/lib/howl/init.lua:64: ../lib/howl/io/process.moon:57: missing declaration for symbol 'sig_IOT'

** (howl:64128): CRITICAL **: Failed to run script: /Users/boonie/Projects/howl/lib/howl/init.lua:144: /Users/boonie/Projects/howl/lib/howl/init.lua:64: /Users/boonie/Projects/howl/lib/howl/init.lua:64: ../lib/howl/io/process.moon:57: missing declaration for symbol 'sig_IOT'
```

**Comment by kirbyfan64** [comment]
@bunster Go back to `process.moon` and change:

```moonscript
  'PWR', 'SYS', 'IOT', 'EMT'
```

back to:

```moonscript
  'PWR', 'SYS'
```

(Sorry if it seems like I'm sending you in circles here! It's kind of hard to fix a problem remotely... :)

**Comment by bunster** [comment]
Yay! It loads up now :) But... There's still one more error:
```
Failed to load application icon: ../lib/ljglibs/gtk/window.moon:96: Couldn't recognise the image file format for file '/Users/boonie/Projects/howl/share/icons/hicolor/scalable/apps/howl.svg' (code: 3)

(howl:65942): GLib-GObject-WARNING **: The property GtkAlignment:bottom-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.

(howl:65942): GLib-GObject-WARNING **: The property GtkAlignment:right-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.

(howl:65942): GLib-GObject-WARNING **: The property GtkAlignment:top-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.

(howl:65942): GLib-GObject-WARNING **: The property GtkAlignment:left-padding is deprecated and shouldn't be used anymore. It will be removed in a future version.
```

**Comment by kirbyfan64** [comment]
@bunster Well, actually, I've had the same exact error since a short time ago...so that's probably a separate bug. Other than that, you now have Howl on OSX!!! :fireworks:

**Comment by nilnor** [comment]
Great to hear you got it up and running! Care to furnish us with a screenshot of the glory? :) It would be interesting to hear more on how it runs, e.g. one thing that comes to mind is key mappings and how they work out.

I had a look around for the icon error, and it seems it has to do with `librsvg`. Do you have that installed from homebrew? If not, try installing that and see what happens. If you have it already, I would suggest seeing if you have an old version and can update from homebrew. There used to be some extra steps required even with it installed it seems, but from what I can determine this should now be handled automatically in the latest homebrew version. Some useful links I found for reference:

https://www.guiguan.net/fix-python-graph-tool-couldnt-recognise-svg-problem/
https://github.com/Homebrew/homebrew/issues/13786

**Comment by shalabhc** [comment]
Woohoo! - this is super exciting :tada:

Great work @bunster, @kirbyfan64, @nilnor!

And yes, I'm interested in the screenshots too :D

**Comment by bunster** [comment]
I didn't have `librsvg` installed via homebrew. Thanks, @nildor, that seems to have solved the startup error :) The icon I see when switching to it though is a terminal with 'exec' written in it.

Here's a screenshot!

![Howl Screenshot](http://i.imgur.com/tu25Cv8.png?1)

**Comment by kirbyfan64** [comment]
Wait...isn't the font supposed to be monospace?? :/

**Comment by shalabhc** [comment]
@bunster - sweet!

The default font may not be available on OS X. Try running `set font=monaco`

**Comment by nilnor** [comment]
Yes, the default font is not available, which is really not all that surprising (it's currently `Liberation Mono`). However, it's possible to specify a list of fonts as I recently discovered, so we could and should update the default. @bunster, good fonts on OS X? We have `monaco` as @shalabhc suggested, but from what I understand it has been replaced as the default sans font by `menlo`. If you could try it out we can update the default to fall back on a good font for OS X.

**Comment by kirbyfan64** [comment]
I've committed all the changes so far in [my fork on the `osx` branch](https://github.com/kirbyfan64/howl/tree/osx). All that's left is to find a good default font (for know I put `monaco`) and figure out the icon not showing up correctly as @bunster mentioned.

**Comment by shalabhc** [comment]
@kirbyfan64 - would also be nice to use `md5` rather than `md5sum` on OSX because the former ships with OSX. https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man1/md5.1.html

**Comment by bunster** [comment]
Sorry, I'm just reading through the documentation on how to do all this stuff.. It seems alt+x isn't working for me. It's entering a symbol instead of opening a command prompt.

**Comment by kirbyfan64** [comment]
@shalabhc Ah, forgot about that one.

It's also worth noting that the Atom text editor uses [`menlo` on OSX](https://github.com/atom/atom/blob/master/static/text-editor-light.less#L7).

**Comment by kirbyfan64** [comment]
@bunster What symbol? Can you take a screenshot?

**Comment by bunster** [comment]
@kirbyfan64 This is the symbol I get when I use ALT+X: ≈

As for a font, it seems `Monaco` is the default OSX version. Though personally, I enjoy using `PT Mono`.

**Comment by kirbyfan64** [comment]
Ah, it seems that the [`alt` key is used on OSX to generate special characters.](https://en.wikipedia.org/wiki/Alt_key#Mac_OS_X)

**Comment by shalabhc** [comment]
@bunster - try replacing 'alt_x' with 'ctrl_x' in lib/howl/keymap.moon. Rebuild, restart and then see if Ctrl+x works? That's the most important shortcut :)

The next step would be to run the `describe-key` command and then try out the usual modifier keys to see what keystroke name they produce.

**Comment by bunster** [comment]
@shalabhc Doesn't seem like that's working. I changed it and then did a make install once more, but it isn't opening a command line.

**Comment by kirbyfan64** [comment]
@bunster So pressing Control+X does nothing? What about pressing Command+X?

**Comment by shalabhc** [comment]
I assume you mean 'make' not 'make install'?

How about this - put the following line at the bottom of your `~/.howl/init.moon` file:

```
howl.signal.connect 'app-ready', howl.command.describe_key
```

Then restart howl - this should drop you directly into the describe key command that captures and displays all keys you press. Then adjust keymap.moon accordingly.

**Comment by kirbyfan64** [comment]
I think Howl should use the Command key on OSX, since it's the one that's usually used for things like `ctrl_c`.

**Comment by bunster** [comment]
None of it is working! :broken_heart: Adding that line to init.moon (which wasn't there, I had to create it), gave this error in the bottom when Howl started up:

![howl startup error](http://i.imgur.com/846uaqt.png)

Yes, I realise my moonscript also has an error in it.. Let's not look at that.

**Comment by shalabhc** [comment]
@bunster - please try running howl without arguments - don't pass a filename to howl.

**Comment by bunster** [comment]
@shalabhc that worked, it said the key I typed was alt_≈ (or alt_16785992). So I changed the keymap file thusly and now the command line is opening!

**Comment by kirbyfan64** [comment]
@bunster What happens if you enter a key like `ctrl_c` into `describe-key`?

**Comment by bunster** [comment]
@kirbyfan64 `ctrl_c` copies the highlighted string.

**Comment by kirbyfan64** [comment]
@bunster Could you do one more key: Command+C?

Ok...so:

1. The `ctrl_*` keys should be mapped to Command. Or maybe they should be equivalent? I know that most OSX editors use Command+C to copy, Command+V to paste, etc.
2. The `alt` key is giving what looks like the Unicode character code to Howl, which is screwing up the key bindings.

**Comment by bunster** [comment]
@kirbyfan64 Command (or `meta`) shows up on its own, but doing `meta+c` simply shows up `key_name = "c"` / `key_code = 99`.

**Comment by kirbyfan64** [comment]
@bunster Could you try just *one last thing*? Go to `lib/howl/bindings.moon`, line 135 and add:

```
  _G.require('moon').p event if _G.require
```

(Again, the whitespace is important!)

Now, open up Howl and try out some keys like Command+C, Meta+X, and Alt+X, and post the output here.

**Comment by bunster** [comment]
@kirbyfan64 it looks the same. This is what it spit out:
For Command/Meta+C
```
{
    [alt] = false
    [shift] = false
    [key_name] = "metaL"
    [meta] = false
    [control] = false
    [key_code] = 65511
    [super] = false
}

{
    [alt] = false
    [shift] = false
    [meta] = true
    [character] = "c"
    [key_name] = "c"
    [control] = false
    [key_code] = 99
    [super] = false
}
```
For Alt+X it was the same as above, excent just an "x" instead.
For Alt+X:
```
{
    [alt] = false
    [shift] = false
    [key_name] = "altL"
    [meta] = false
    [control] = false
    [key_code] = 65513
    [super] = false
}

{
    [alt] = true
    [shift] = false
    [meta] = false
    [character] = "≈"
    [key_name] = "u+2248"
    [control] = false
    [key_code] = 16785992
    [super] = false
}
```

**Comment by kirbyfan64** [comment]
@bunster Command+C and Meta+C have the exact same output? That's pretty weird.

**Comment by kirbyfan64** [comment]
The weird behavior of alt is because of [this](https://en.wikipedia.org/wiki/Alt_key#Mac_OS_X). OSX is giving Howl the Unicode key character. *sigh* This is more complicated than it looks.

I'm not sure what to do about Command and Meta, though. If they really do both come up with `meta` as `true`, the result is going to suck however it's done. I'm guessing Meta could be an alias for Alt, and Howl could have table of the Alt Unicode characters -> the original key. Ideally, Command would be an alias (or a replacement) for Control, but if it shows up as Meta, it will be kind of weird...

**Comment by bunster** [comment]
@kirbyfan64 They're the same key, look [here](http://i.imgur.com/HhDxWSV.png).

**Comment by kirbyfan64** [comment]
@bunster Ah, right. I feel stupid right now. :/

**Comment by bunster** [comment]
@kirbyfan64 :( You're not stupid... Just unique ;)

**Comment by shalabhc** [comment]
@bunster 
> Command (or meta) shows up on its own, but doing meta+c simply shows up key_name = "c" / key_code = 99. 

Probably because there is no special handling for the 'meta' modifier here:
https://github.com/nilnor/howl/blob/master/lib/howl/bindings.moon#L134

Should be easily fixable, by doing for 'meta' what is done for 'alt' in that function.

The Alt key isn't really used much on Macs so I'm not too worried about it for now. In fact users will be used to inserting accented characters with it and it should just work identically in Howl.

Regarding porting the keymap, I think the right approach is not to map a modifier key (e.g. all `alt_<key>` on Linux become `meta_<key>` on OSX), but to have a distinct keymap.moon for OSX. This gives greater control and allows better consistency with the target platform.

**Comment by kirbyfan64** [comment]
@shalabhc I have handling of meta on the osx branch but haven't committed it yet. I like your idea of separate keymaps...but that has its own problems. Namely, what would happen for OSX users if a bundle author used Alt as the base key in a custom keybinding.

Since Command/Meta on Macs is largely used where Control is on Windows/Linux, a possible solution would be to map Control to Alt and Meta to Control on Macs (though that may end up slightly confusing... :/ ).

Next time someone tells me about how much more "sane" OSX is in comparison to Linux or Windows, I'm going to point them to this issue.

**Comment by kirbyfan64** [comment]
(It might be worth looking into what other editors such as Sublime and Atom do in this scenario.)

**Comment by shalabhc** [comment]
@kirbyfan64 - good point regarding bundle keymaps. Perhaps bundles could provide multiple keymaps and one gets activated based on a config setting that also choses the global keymap. In any case, users will be able to customize key bindings on OSX for any bundle command, if the bundle does not provide multiple keymaps.

Atom seems to provide multiple distinct keymaps - they even have one called 'emacs': https://github.com/atom/atom/tree/master/keymaps

@bunster - it would be great if you could document the steps needed to build Howl on OSX.

**Comment by nilnor** [comment]
Re key maps:

One alternative would be to have a synthetic modifier key, say `action`. For OS X the command key (meta) would be mapped to `action` as well as to `meta` (and possible also to `cmd`). For other OS:s the control key would be mapped to `action` as well as to `ctrl`. This would allow for generic key bindings such as `action_c` that would be triggered by `ctrl_c` on "normal" platforms and `cmd_c` on OS X, while still allowing for explicit key bindings such as `ctrl_x` or `cmd_x`.

Alt bindings are still complicated as they just don't work properly on mac, so any alt binding would need to have explicit counter parts in keymaps. E.g. `alt_x` needs a corresponding `cmd_X` or something similar.

Thoughts?

@kirbyfan64 Could you open a pull request with the changes so far, so we can look at getting the needed stuff into master?

**Comment by shalabhc** [comment]
@nilnor - Adding a synthetic modifier `action` seems reasonable. This could cover most of the common cases. For special cases you'd still need an OSX specific keymap. Perhaps there could be a default keymap and platform specific override files. This lets us specify at least the common shortcuts in the default keymap using `action` and override only when needed.

**Comment by kirbyfan64** [comment]
@nilnor That's a great idea!

What if there were two "virtual" keys like this: `action` and `second?` (Yeah, the name sucks; I can't think of anything else right now.)

On Linux:
- `action` is `ctrl`
- `second` is `alt`

On OSX:
- `action` is `meta`/command
- `second` is `ctrl`

From what I've seen, that's how the keys usually get used.

**Comment by kirbyfan64** [comment]
Today I asked someone who uses a Mac, and he had a similar conclusion: Linux Control is like OSX Meta, and Linux Alt is like OSX Control. So what I said in my last comment should be about right (with better names, of course!).

**Comment by nilnor** [comment]
@kirbyfan64 I'm not sure about having two synthetic modifiers. What you say makes sense, but I'm concerned that it all could become confusing, with everyone unsure exactly what a particular key binding actually is triggered by. That being said, that just means I'm currently unsure, not that it's a bad idea - as you say `second` is a bad name and perhaps a better name would make it sound more appealing :) How about `opt` (which coincidentally is what is used for the alt key on macs, or used to be at least)?

At any rate, it's time to release 0.3 now, so the binding rework will have to wait until the release is cut. We'll have ample time to sort out everything properly after that, and hopefully also be sure Howl works as it should in general on OS X.

**Comment by kirbyfan64** [comment]
`opt` sounds good!

**Comment by nilnor** [comment]
Also, regarding the application icon I don't think this is something that can fixed by code. Rather it seems to be something that would be done by creating a proper application bundle for Howl on OS X (see https://wiki.gnome.org/Projects/GTK+/OSX/Bundling for more information).

**Comment by shalabhc** [comment]
Using `opt` to mean anything other than the Option key on a Mac will be confusing. IIUC, @kirbyfan64  proposed the second synthetic modifier be mapped to the Control key on Macs.

I think it's useful to think about shortcut usability from the user's point of view - I can think of two kinds of users:

- Mac only users: for these users the best shortcut keys are the ones that are most consistent with other Mac applications and the published human interface guideline (https://developer.apple.com/library/mac/documentation/UserExperience/Conceptual/OSXHIGuidelines/Keyboard.html).

- Cross platform users: these users might  want Howl to behave identically when keys in the same *position* are used on either Linux or OSX. I (and I believe most people) memorize shortcut modifiers by their location.

I'm not sure what the best solution for this is, but just wanted to put forth these use cases.

**Comment by nilnor** [comment]
@shalabhc Yes, I realized later that I've made a mental mix-up when it came to `opt`, since as you say it actually is the alt key there. So, strike that suggestion. The two use cases are good, and points to different solutions. I believe we have to cater at least in part to the first case - not having `cmd+o` map to the open command seems like it would just be weird and counter-intuitive on mac. Whether aiming for full compatibility with HIG is a good idea is another question though. I'm not sure about the cross-platform position argument though. If people switch between the OS:s, they'll probably switch their expectations of what keys trigger a copy or paste as well. Considering all the other apps would then work that way.

**Comment by shalabhc** [comment]
@nilnor - I agree we should at least cater to the first case - that would constitute 'proper' OSX support. I think 100% compatibility with the HIG  doesn't make much sense for Howl which is a cross platform app with a non-conforming UI, but we should definitely see and use the recommended shortcuts where they make sense. 

I think the second case also has utility, and should be very simple to implement. Remembering the small set of shortcuts for common actions such as copy/paste might be easy for a user, but remembering two entire sets of all shortcuts in Howl is probably not that easy. When I used emacs on a Mac, it was nice that all my finger memory from Linux just transferred over (and it was also irritating that the copy/paste shortcuts were not consistent, but I couldn't have both).

**Comment by nilnor** [comment]
@kirbyfan64 Now with 0.3 released, do you want to open a PR for the compilation fixes?

**Comment by JorjBauer** [comment]
The discussion about keyboard bindings is largely a matter of expectation of an app to behave a certain way, and wrapping in GTK, that expectation is violated.

Seems to me that having a curses variant that one could run in Terminal (or iTerm, or cygwin, or pick-your-terminal) would correct a lot of those expectations, and simplify the codebase.

**Comment by JaredSartin** [comment]
Unsure if it would simplify the code base, but it would mean it gets more
modular. Unfortunately, it will take a bit of work to get a curses front
end.

I would love to see it happen, don't get me wrong, but it's a decent size
rewrite (unless the underlying engine is written cleanly).

**Comment by nilnor** [comment]
@JorjBauer well using Gtk does mean that Howl its unlikely to provide a fully native feel, but that does not mean that we shouldn't make everything as straightforward as possible. Having key bindings work in a sane way on OS X make sense to me, no need to make it worse than it has to be.

As @JaredSartin says, producing a curses variant is not a small undertaking. It would if anything complicate the code base due to the need for abstracting away the interface parts, etc. Not to say it wouldn't be very nice to have some day, but someone would have to put in a large amount of effort.

**Comment by nilnor** [comment]
The `uni-bindings` branch now contain support for a synthetic `action` modifier (pr #95). As I don't have a mac lying around someone would have to check that out and see how it works in practice.

**Comment by nilnor** [comment]
For the record, build fixes and a default keymap is now in place in master. What's missing for closing this issue is installation instructions for OS X.

**Comment by shalabhc** [comment]
Quick update - I built Howl and OS X and it was incredibly easy:

1. Install homebrew (http://brew.sh)

2. Install `gtk+3` and `wget` via homebrew:

```
brew install wget
brew install gtk+3
```

3. Build Howl as usual:

```
git clone https://github.com/howl-editor/howl
cd howl
make -C src
```

Now `./src/howl` can be executed. This is not a proper OS X application bundle, but it runs.

What doesn't work well:
- Font Awesome icons dont work as it appears custom fonts are not loaded. The workaround is to manually install Font Awesome via Font Book and change the definition code to refer to 'FontAwesome' instead of 'Font Awesome'.
- Some line height calculations seem off as after changing font or font size:
  - the bottom of the command line text appears clipped occasionally: 
![image](https://cloud.githubusercontent.com/assets/7066873/11454503/c00bfd7c-95e2-11e5-9009-59351b0f5f7e.png)
 
  - the view titlebar is clipped to the left and right:
![image](https://cloud.githubusercontent.com/assets/7066873/11454490/858de7dc-95e2-11e5-88a0-81d21a424834.png)

- The `run` binding of `ctrl_meta_x` is too cumbersome for a commonly used shortcut. Maybe this should be just `meta_e` (for execute).
